pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -674,6 +847,34 @@ func (c *Controller) constructWorkload(pod *v1.Pod, waypoint *workloadapi.Gatewa
 	for _, podIP := range pod.Status.PodIPs {
 		addresses = append(addresses, parseIP(podIP.IP))
 	}
+	for _, svc := range c.ambientIndex.servicesMap {
+
+		if svc.Attributes.ServiceEntry == nil {
+			// if we are here then this is dev error
+			log.Warn("dev error: service entry spec is nil; it should have been populated by the service entry handler")
+			continue
+		}
+
+		if svc.Attributes.ServiceEntry.WorkloadSelector == nil {
+			// nothing to do. we construct the ztunnel config if `endpoints` are provided in the service entry handler
+			continue
+		}
+
+		if svc.Attributes.ServiceEntry.Endpoints != nil {
+			// it is an error to provide both `endpoints` and `workloadSelector` in a service entry
+			continue
+		}
+
+		sel := svc.Attributes.ServiceEntry.WorkloadSelector.Labels
+		if !labels.Instance(sel).SubsetOf(pod.Labels) {
+			continue
+		}
+
+		vipsToPorts := getVIPsFromServiceEntry(svc, nil)
+		for vip, ports := range vipsToPorts {
+			vips[vip] = ports
+		}
+	}
```

Comment:
this should be DRY'ed with the other implementation (which is more correct, since it has the necessary `vip := c.network.String() + "/" + vip // services must be on our network, don't inherit from workload` lines)

Reply:
addressed in https://github.com/istio/istio/pull/45621/commits/5a0ee021b0906e18cdf5c1c31615dbbc2d15e65f